sdk/core/src/overridable_context.rs

```diff
@@ -0,0 +1,203 @@
+use crate::{Context, TypeMapContext};
+use std::any::{Any, TypeId};
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+#[derive(Debug)]
+enum PreviousContext<'a> {
+    Context(Cow<'a, Context>),
+    OverridableContext(&'a OverridableContext<'a>),
+}
+
+/// Execution context.
+///
+/// The `OverridableContext` is modeled as an overridable chain. In other words, the `OverridableContext` allows to
+/// override any value without losing the original one. The resulting type map is the union of the
+/// type map of every chained `OverridableContext`, giving precedence to the last inserted one. As a simple
+/// example suppose to have a `OverridableContext` with the key value of `Struct1` -> `Hello` and `Struct2` ->
+/// `World`. Let's call it `C1`.
+/// If we chain (using the `from_previous_context` function) this `OverridableContext` we get another
+/// `OverridableContext`, called `C2`.
+/// Calling `get` on C2 is now the same of calling `get` on `C1`: `Struct2` gives `World`. But now
+/// we can replace `Struct2` with `Solar system` without losing the original `C1`'s `World`. If we
+/// do that, `C2.get::<Struct2>()` will give you `Solar system`, while `C1.get::<Struct2>()` will
+/// give you `World`.
+#[derive(Debug)]
+pub struct OverridableContext<'a> {
```

Comment:
I don't really get why we need two types `Context` and `OverrideableContext`. Is this just to ensure that one is immutable and the other isn't? Wouldn't the fact that the `send` method takes an immutable reference guarantee that?